PEP 844: ``public`` and ``private`` builtins (#5079)
python/peps: peps/pep-0844.rst

PEP: 844
Title: ``public`` and ``private`` builtins
Author: Barry Warsaw <barry@python.org>
Discussions-To: Pending
Status: Draft
Type: Standards Track
Created: 05-Aug-2026
Python-Version: 3.16
Post-History: Pending

Abstract
========

This PEP proposes adding two new builtin functions, ``public()`` and ``private()``, which document
the public interface of a module by keeping its ``__all__`` synchronized with the names actually
defined in that module.  Both are used as decorators (``@public`` and ``@private``) on class and
function definitions, so that a name's visibility is declared exactly once, at the point where the
name is defined.  ``public()`` additionally has a function call form for names that cannot be
decorated, such as constants.

For example:

.. code-block:: python

   # spam.py
   @public
   class Public:
       ...

   @private
   class Private:
       ...

   public(SEVEN=7)

> .. code-block:: pycon

Reviewer: I don't think the conference has a syntax. 😉
```suggestion
.. code-block:: python
```

Reviewer 2: It's not "Python conference" but "Python console" :)

> For Python console output or doctests, ...

https://pygments.org/docs/lexers/#pygments.lexers.python.PythonConsoleLexer

> For Python source code (version 3.x).

https://pygments.org/docs/lexers/#pygments.lexers.python.PythonLexer

   >>> import spam
   >>> spam.__all__
   ['Public', 'SEVEN']

The proposed semantics are those of the third-party `atpublic
<https://public.readthedocs.io>`__ package, which has provided this functionality since 2016.

This PEP is an adjunct to :pep:`842` and PEP 843; see `Relationship to PEP 842 and PEP 843`_.

Motivation
==========

> The :attr:`module global variable <module.__all__>` ``__all__`` is the mechanism Python currently

Reviewer 3: This looks weird in the output because "module global variable" is rendered as code. I suggest doing this:

```suggestion
The module global variable :attr:`~module.__all__` is the mechanism Python currently
```

``@private`` deliberately does not create an empty ``__all__``, because doing so would silently
change the meaning of ``from spam import *``.  With no ``__all__``, a wildcard import binds every
name not beginning with an underscore; with ``__all__ = []`` it binds nothing.  A decorator whose
purpose is documentation should not have that effect.

It follows that ``@private`` alone does not exclude a name from ``from spam import *``.  Excluding
names is the job of ``@public``: as soon as any name in the module is marked public, ``__all__``
exists, and everything not marked public is excluded automatically.  ``@private`` records the
author's intent; ``@public`` is what makes that intent observable.

.. note::

   ``private()`` does *not* support a function call form, as no valid use case for it has been
   identified or requested by users of the ``atpublic`` module.  See `Open Issues`_ for
   further discussion.

Restrictions
------------

Note that because ``@public`` is primarily used to keep the ``__all__`` module global in sync, only
module-level objects may be decorated.  Decorating a method inside a class body is not supported,

> since ``__all__`` documents module contents, not class contents.

Reviewer 3: This is very nitpicky, but what happens if you decorate a method? Is it an error, or does it append the name to `__all__`?

Author: In the current implementation, it will append the name to `__all__`.  The implementation does not inspect its calling scope to confirm it's at module scope.  I think doing so wouldn't be worth it, given the admonitions I'm adding to the PEP in the next draft.  It's something I'll make clear in the documentation.

Because ``__all__`` must be mutable for these functions to append to it, a module that assigns
``__all__`` itself must assign a list.  A module that wants an immutable ``__all__`` can freeze it
after the last declaration with ``__all__ = tuple(__all__)``.

Rationale
=========

Why builtins?
-------------

The declaration of a module's public interface is a fundamental and often requested property,
especially as code bases grow.  Many use cases have been identified in discussions, and different
approaches have been developed in different libraries and applications.  Enough experience has been
gained over the decade of ``atpublic``'s existence that requiring a third-party dependency (or an
import at the top of every module) to spell something this fundamental is friction that discourages
its use.  It is also awkward in exactly the places where it matters most: the standard library

> itself, and small single file modules.

Reviewer 3: This needs a hyphen:

```suggestion
itself, and small single-file modules.
```

assignment, including ``__all__ += [...]`` and ``__all__.append(...)``, precisely because real code
does this.  They already special-case namespace-creating callables whose behavior is not evident
from the grammar, such as :func:`~collections.namedtuple`, :class:`~typing.TypedDict`, and
:func:`~dataclasses.dataclass`.  Adding ``public()`` to that list is an increment on work these
tools have already done, not a new category of problem.

If this PEP is accepted, that support is expected to follow quickly, for the ordinary reason that
tools support what the language provides.  In the interim (and for older tool versions) the return
value of ``public()`` gives an entirely explicit spelling that requires no special support at all:

.. code-block:: python

   SEVEN = public(SEVEN=7)

Here the binding is a plain assignment, visible to every tool that parses Python.  This form is a
transition aid rather than the recommended spelling, and it should not be needed for long.

The conclusion is that the data and type alias use cases, which are the places a decorator
genuinely cannot be utilized, do not require new syntax at all.  They require a function call that

> tools can learn to read, alleviating the need for a dedicated, new ``export`` keyword.

Reviewer 3: "Alleviating" feels awkward here; you alleviate a problem, not a solution. Maybe "removing" instead?

```suggestion
tools can learn to read, removing the need for a dedicated, new ``export`` keyword.
```

Author: Thanks!  Rewritten in a different way.

> .. _pep-844-urgency:

Reviewer 3: This label is unused.

Author: Since labels are project-wide, this could still be useful.

   implementation.  This is a problem about *where you add the declaration*.

2. **Runtime consequences.** ``__all__`` declares the public API, but the only place that
   declaration is enforced is ``from spam import *``.  It has no effect on attribute access,
   :func:`dir`, :func:`help`, or autocompletion, so a name that is intended to be kept private is
   indistinguishable from public names to these patterns of module introspection.  This is a
   problem about *what the declaration does*.

This PEP addresses only the first.  It takes the position that the first problem is the more
pressing and the more broadly applicable of the two, that it can be solved without new syntax and
without a new variable, and that solving it does not commit Python to any particular answer to the
second.

.. _pep-844-why-all:

Why ``__all__`` and not ``__export__``
--------------------------------------

:pep:`842` proposes a new ``__export__`` variable.  This PEP proposes to keep using ``__all__``.

:pep:`842` gives two reasons why ``__all__`` is inadequate.  The first is that ``__all__`` drifts
out of sync with the module.  That is true, and it is precisely the problem ``atpublic`` and this

> PEP solves.  However a *new list of string literals in the same distant part of the file* does not

Reviewer 3: 1. The subject here is plural, so it should be "solve" instead of "solves".
2. "However" needs a trailing comma.
```suggestion
PEP solve.  However, a *new list of string literals in the same distant part of the file* does not
```

``__all__``.  A module conscientious enough to maintain ``__export__`` accurately would have been
conscientious enough to maintain ``__all__`` accurately; the ones that drift will drift in both.

This PEP takes no position on whether unexported-name warnings are desirable.  It observes only that
the bookkeeping question is separable from the runtime-semantics question, and it answers the
former.  ``public()`` populates a list; if Python later decides that some list should carry runtime
consequences, ``public()`` can populate that one instead, or both.  Nothing here closes the door on
:pep:`842`.

Why PEP 843 is a good companion
-------------------------------

This PEP does **not** solve the DRY problem for re-exports, and cannot do so gracefully.  A "hub
module" that pulls names out of private submodules must currently write each name three times:

.. code-block:: python

   from ._core import Widget
   public(Widget=Widget)

``Widget`` is named once to import it, and twice more to export it.  That's a big violation of DRY!
Hand-maintaining ``__all__`` would name it only twice, so for re-exports specifically, ``public()``

> is not merely unhelpful, it is a step backwards.

Reviewer 3: Comma splice:

```suggestion
is not merely unhelpful; it is a step backwards.
```

or:

```suggestion
is not merely unhelpful, but rather it is a step backwards.
```

Author: Thanks!  Rewritten in a different way.

> PEP: 844

Reviewer 2: And add to the CODEOWNERS file.

Author: Thanks!  I always forget that one.

Title: ``public`` and ``private`` builtins
Author: Barry Warsaw <barry@python.org>
Discussions-To: Pending
Status: Draft
Type: Standards Track
Created: 05-Aug-2026
Python-Version: 3.16
Post-History: Pending

Abstract
========

This PEP proposes adding two new builtin functions, ``public()`` and ``private()``, which document
the public interface of a module by keeping its ``__all__`` synchronized with the names actually
defined in that module.  Both are used as decorators (``@public`` and ``@private``) on class and
function definitions, so that a name's visibility is declared exactly once, at the point where the
name is defined.  ``public()`` additionally has a function call form for names that cannot be
decorated, such as constants.

For example:

.. code-block:: python

   # spam.py
   @public
   class Public:
       ...

   @private
   class Private:
       ...

   public(SEVEN=7)

.. code-block:: pycon

   >>> import spam
   >>> spam.__all__
   ['Public', 'SEVEN']

> The proposed semantics are those of the third party `atpublic

Reviewer 2: ```suggestion
The proposed semantics are those of the third-party `atpublic
```

<https://public.readthedocs.io>`__ package, which has provided this functionality since 2016.

This PEP is an adjunct to :pep:`842` and PEP 843; see `Relationship to PEP 842 and PEP 843`_.

Motivation
==========

> The `module global variable <https://docs.python.org/3/reference/simple_stmts.html#module.__all__>`_

Reviewer 2: ```suggestion
The :attr:`module global variable <module.__all__>`
```

often far from the objects whose names are contained in it.  An object defined at one point in the
file is repeated as a string literal in an ``__all__`` list somewhere else, usually at the top of
the file.

Nothing keeps the two in sync, leading to these problems:

* Names get added to the module but never added to ``__all__``.
* Names get removed from or renamed in the module but not in ``__all__``, so ``from spam import *``
  raises :exc:`AttributeError`.
* It's easy to typo a name (or leave out a list item delimiting comma) in ``__all__``.
* Drift is only detectable in one direction.  A linter can flag a name in ``__all__`` that doesn't
  exist as an object in the module, but no tool can flag a public name missing from ``__all__``,
  because nothing in the source says that name was meant to be public.
* Readers of the code must scroll to a different part of the file (or a different screen) to answer
  "is this name public?"

The convention of prefixing private names with an underscore addresses a related but different
problem, and :pep:`842` describes at length why :pep:`prefixing is not by itself a sufficient

> answer <842#prefixed-names-aren-t-necessarily-a-great-solution>`.

Reviewer 3: Would it help if I added a label to this section (or any section you want to reference) in PEP 842? This could break if I change the header for some reason.

Author: Yes, that would be great!

Reviewer 3: Done, this section is under the label `pep-842-why-not-prefixed-names`:

```suggestion
problem, and :pep:`842` describes at length why :pep:`prefixing is not by itself a sufficient
answer <pep-842-why-not-prefixed-names>`.
```

Let me know if there are any other sections you want to link to.

Author: > Let me know if there are any other sections you want to link to.

Thanks!

The pattern proposed here -- declaring visibility at the definition site with a decorator -- is not
new or speculative.  The ``atpublic`` package on PyPI has implemented it for a decade, and it is
already depended on by a number of projects.  What this PEP proposes is that the pattern is common

> enough, and useful enough, to be spelled without a third party dependency.  Thus it proposes to add

Reviewer 2: ```suggestion
enough, and useful enough, to be spelled without a third-party dependency.  Thus it proposes to add
```

author's intent; ``@public`` is what makes that intent observable.

.. note::

   ``private()`` does *not* support a function call form, as no valid use case for it has been
   identified or requested by users of the ``atpublic`` module.  See `Open Issues`_ for
   further discussion.

Restrictions
------------

Note that because ``@public`` is primarily used to keep the ``__all__`` module global in sync, only
module level objects may be decorated.  Decorating a method inside a class body is not supported,
since ``__all__`` documents module contents, not class contents.

Because ``__all__`` must be mutable for these functions to append to it, a module that assigns
``__all__`` itself must assign a list.  A module that wants an immutable ``__all__`` can freeze it
after the last declaration with ``__all__ = tuple(__all__)``.

Rationale
=========

Why builtins?
-------------

The declaration of a module's public interface is a fundamental and often requested property,
especially as code bases grow.  Many use cases have been identified in discussions, and different
approaches have been developed in different libraries and applications.  Enough experience has been

> gained over the decade of ``atpublic``'s existence that requiring a third party dependency (or an

Reviewer 2: ```suggestion
gained over the decade of ``atpublic``'s existence that requiring a third-party dependency (or an
```

The conclusion is that the data and type alias use cases, which are the places a decorator
genuinely cannot be utilized, do not require new syntax at all.  They require a function call that
tools can learn to read, alleviating the need for a dedicated, new ``export`` keyword.

.. _pep-844-urgency:

Is this urgent?
---------------

`Guido van Rossum <https://discuss.python.org/t/108353/46>`__ raised this question about :pep:`842`,
and it applies with equal force here:

   But Python has existed without this feature for over 35 years -- is
   it really urgent?  Remember the Zen of Python, which says "Now is
   better than never.  Although never is often better than *right*
   now."

No.  This PEP is not urgent, and it does not claim to be.  Nothing about module name visibility, or
about a module's exported public API, is urgent.  But urgency is the wrong test to apply to this
particular proposal, for three reasons.

**The feature is not new.** This PEP does not ask Python to adopt an untried idea; ``atpublic`` has
implemented these exact semantics since 2016.  The question is not "should Python have this?" since

> users who want it already have it, but "should having it cost a third party dependency?"  A decade

Reviewer 2: ```suggestion
users who want it already have it, but "should having it cost a third-party dependency?"  A decade
```

``__export__`` if absent, and call ``list.append`` -- the same operations, expressed in bytecode
rather than a call.  The saving is the function call dispatch, not the underlying work.  That is a
real difference, but it is a constant factor on an already small constant, not a difference in kind.

**A C implementation is feasible and fast.** This is the point on which a builtin is strictly better
positioned than the third-party package.  ``atpublic`` shipped a C implementation of ``public()``
for a time, and it was substantially faster than the pure Python version.  It was ultimately
dropped, not because it did not work, but because requiring a compiled extension module in a third
party package is a significant packaging and installation burden for a library this small -- a
burden borne entirely so that the pure Python fallback could be avoided.

That trade-off does not exist in CPython.  A builtin is compiled as part of the interpreter, so the
fast implementation is simply *the* implementation, with no wheel platform support matrix, no
fallback path, and no optional extra.  Moreover, a C implementation inside the interpreter can do

> less work than any third party one: the decorator form can access the calling frame's globals

Reviewer 2: ```suggestion
less work than any third-party one: the decorator form can access the calling frame's globals
```

Reference Implementation
========================

The `atpublic <https://public.readthedocs.io>`__ package, available on PyPI and maintained since
2016, implements the proposed semantics in pure Python.  Its `source repository
<https://gitlab.com/flufl/public>`__ is hosted on GitLab.

A CPython implementation has not yet been written.

For a time, ``atpublic`` also included a C implementation of ``public()``, which was considerably
faster than the pure Python one.  It was dropped for packaging reasons that do not apply to a
builtin.  See :ref:`pep-844-performance`.

One divergence is worth noting.  ``atpublic`` 7.0.0 and earlier create ``__all__`` in the
``@private`` case, contrary to the specification above.  This was identified as a bug while drafting
this PEP, and will be corrected in ``atpublic`` 8.0.0, which is in pre-release at the time of this
writing.

Rejected Ideas
==============

New ``export`` syntax instead of decorators
-------------------------------------------

:pep:`842`, in its current revision, proposes an ``export`` soft keyword covering the same ground as
this PEP -- ``export def``, ``export class``, ``export NAME = value``.  Its `Rejected Ideas

> <https://pep-previews--5077.org.readthedocs.build/pep-0842/index.html#rejected-ideas>`_ section

Reviewer 2: ```suggestion
:pep:`Rejected Ideas <842#rejected-ideas>` section
```

  any of them can sensibly be accepted.
* Should ``populate_all()``, ``atpublic``'s heuristic "infer ``__all__`` from what's defined here"
  function, also be included?  This is deferred for now; a heuristic is a harder case to make for a
  builtin than the two explicit declarations are, and is less essential for improving module
  visibility ergonomics.
* Should ``private()`` support a function call form, for symmetry?  ``atpublic`` does not provide
  one and no need for it has ever been demonstrated or requested.
* Should the standard library itself adopt these decorators, and if so on what schedule?  This
  question is entangled with :ref:`pep-844-performance` and should be settled with startup
  measurements in hand.  Also, as with all new capabilities (such as lazy imports), Python's policy
  is generally not to wholesale update the stdlib to embrace the new functionality.  These new
  functions can be utilized opportunistically in modules where the most benefit can be gained, or
  when a module undergoes substantial rewrite.
* Import time benchmarks for the C implementation are outstanding.

Acknowledgements
================

> Thanks to Peter Bierma and Neil Girdhar, whose :pep:`842` and PEP 843 prompted this proposal, and to

Reviewer 4: Even if PEP 843 PR is not merged yet, it would be better to use a link:

```suggestion
Thanks to Peter Bierma and Neil Girdhar, whose :pep:`842` and :pep:`843` prompted this proposal, and to
```

Author: I can't apply this one until 843 is actually published, otherwise the build fails.  I have a TODO to clean these up.

Reviewer 4: Oh, I didn't know that it emits an error. So it makes sense to not add a link yet!
